ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -493,6 +505,46 @@ public boolean exitMaintenanceMode() {
     return exitMaintenanceMode;
   }
 
+  /**
+   * Maintenance record. This is an internal structure in helix. It might be changed with new version of helix controller.
+   */
+  static class MaintenanceRecord {
+    @JsonProperty("DATE")
+    public String date;
+    @JsonProperty("TIMESTAMP")
+    public String timestamp;
+    @JsonProperty("OPERATION_TYPE")
+    public String operationType;
+    @JsonProperty("REASON")
+    public String reason;
+    @JsonProperty("TRIGGERED_BY")
+    public String triggeredBy;
+    @JsonProperty("USER")
+    public String user;
+
+    public static final String OPERATION_TYPE_EXIT = "EXIT";
+    public static final String OPERATION_TYPE_ENTER = "ENTER";
```

Comment:
Just realized that Maintenance operation type can't be used to determine if a cluster is in MM or not. Enable/Disable MM takes two steps, first step is to send a maintenance signal, which would create a znode at /ClusterName/CONTROLLER/MAINTENANCE. The second step is to create an entry in HISTORY. 
Upon receive the maintenance signal, the cluster will be put in to MM. 

In this way, if the process, who sent the signal, dies after the signal is sent, then we won't be able to see the exit history.

Anyway, we don't even have to check the operation type in the last maintenance history entry. If the reason is different, we should not exit MM.

Reply:
In the above procedure, when do we write the ENTER history?

I saw two steps: 
write: /ClusterName/CONTROLLER/MAINTENANCE
write: EXIT HISTORY

Reply:
Enter and Exit history are the same thing, it's just a record to insert to the maintenance history list. When we enter MM, helix would create  /ClusterName/CONTROLLER/MAINTENANCE and after that it will write an entry to /ClusterName/CONTROLLER/HISTORY.